### PR TITLE
Relational Interfaces Rework 🐰🕳️

### LINKED ISSUE #3834: Fix O2M treating new items as existing if they have a manual primary key
Directus v9.0.0-rc32

When trying to setup a O2M field, and updating it, the added item is not saved (see gif below).
I want to setup for my Amenity collection a field that will rely on another collection to record the "feedback criteria" I want.
I've setup the feedback criteria collection, then in my amenity collection I've setup the O2M relationship.
But when then I want to add feedback criteria to an amenity, it does not work.
Is it maybe because of the translation field in the feedback critera collection?

### LINKED ISSUE #3889: Add pagination to relational interfaces
If you have > 10 languages, the list gets too long. We should truncate after the first X, and have a "show all" button

This also extends to the repeater interface

### LINKED ISSUE #5540: Use SEARCH instead of GET for o2m relational interfaces
Currently, the o2m interface will fetch and render the first 100 items inline. If you are relating to a collection that uses UUIDs for IDs, there's a change that the full URL starts to grow too big for certain browsers/webservers to handle, resulting in a `414 request-URI too large` error (see https://github.com/directus/directus/discussions/5419).

To resolve this, we should replace this call:

https://github.com/directus/directus/blob/bdbe28508934d97900c24cf2dd7f66c600893250/app/src/interfaces/list-o2m/list-o2m.vue#L307-L312

with [a SEARCH request](https://docs.directus.io/reference/api/introduction/#search-http-method) where the query is in the body

### LINKED ISSUE #5752: Changing one item in O2M corrupting data in other items in the same O2M relation if you click on them before saving at top level
@MiniDigger thanks for the suggestion to check the display template. In my case, it already has one. I can also verify that when I change one item, save, go into the other item (where it displays the change to the field from the first one), it comes back from the API properly. It's like it believes I'm making the change from the first one in all of them though.

To demonstrate, the attached screenshot shows the API call for the item with ID 22 being correct. Yet the interface is displaying the change I made to ID 21. It actually doesn't matter if, after making the change to the first one, I save it or just click the X. The damage is done. All will be displaying the tentative change if I go in them. Again, if I don't go into any others and save it all the way through to the top level, it will only make the expected change. The oddities occur only after I click on any others prior to saving.

The workaround is having to make changes one at a time, being careful to not click anything other than the one I am changing, saving at the top level each time. Obviously not ideal.

This is a different issue because setting the display template doesn't change anything, and I _think_ my issue actually affects O2M (and perhaps not others). As a result, I'll peel this off into a new issue.

Screenshot 2021-05-18 22 32 04

_Originally posted by @jpetryk in https://github.com/directus/directus/issues/5681#issuecomment-843742689_

### LINKED ISSUE #5996: _ncontains doesn't work on o2m nested dataset
Filter not working properly on nested items (on many to one or one to many)

- Left column: ID
id: 110

- Right column: Jakso (info from another collection)
id: 110 --> 1/2021, 2/2021, 3/2021

Trying to only show the **ID:s** **which does not have 2/2021 anywhere in "Jakso"** (screenshot).

In the screenshot ID:110 has got 2/2021, but is still shown.

image

### LINKED ISSUE #6085: Row reordering changes include entire client cache on request
## Steps to Reproduce

1. Create a `rows` collection, with a `entries` m2a field holding (sorted) items from a `cells` collection.
2. In the cells collection, set up a further nested m2o field called `value` with an item from a `values` collection.
3. In the `cells` display template, show both `cells.id` and `cells.value.id`.
3. Give a restricted user permissions to create/read/update `rows` and `cells`, but only read permissions on `values`.
4. As admin, create two `cells` containing non-null `values` items. Put these cells in a `rows` item.
5. As a restricted user, attempt to re-order.

## Expected and Actual Behavior

I expect to be able to reorder these items, but instead, I get a `Forbidden` exception. Going into the network debugger, I notice the shape of the request is as follows:

```ts
{
  "entries": [
    {
      "0": {
        "$index": 0,
        "entity": "cell",
        "entity_id": {
          "id": 1,
          "value": {
            "id": 1
          }
        },
        "id": 1,
        "sort": 1
      }
    }
    ... etc ...
  ]
}
```

## Environment

This is a followup to #6041, where similar behavior was occurring, but for ostensibly different reasons. I'm using Directus v9.0.0-rc.73 on macOS + npx, using Firefox 88.

### LINKED ISSUE #9990: When editing m2m item the payload contains both junction table aswell as related table object
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes my problem.

### Describe the Bug

When editing an item through the m2m interface, the payload contains (as intended) an object with the PK of the junction table item and the changed values. But this object also contains an object of the related table, who's values haven't changed.
Then the app updates both the item in the junction table aswell as the item in the related table.

_For admins this doesn't matter, it's just an unnecessary update._

But for users with limiting permissions this could result in a **`permission denied`** error if this user isn't granted update rights on the related table. This error prevents saving any of the changes made, so it's quit annoying.

**payload**
image

### To Reproduce

1. Create a collection 'book' with a field called 'title'
2. Deny update rights for this collection
3. Create a collection 'author'
4. Create a m2m relation between 'author' and 'book'
5. Select the `many to many` interface on this relation and use `{{book_id.title}}` as display template
5. Add an extra field 'RevenuePercentage' to the author_book collection
6. Note that when editing the 'RevenuePercentage' of a book author through the m2m interface on the `author` collection the payload also contains a book_id object

### Errors Shown

`[FORBIDDEN] You don't have permission to access this.`

### What version of Directus are you using?

9.0.1

### What version of Node.js are you using?

v16.13.0

### What database are you using?

MSSQL

### What browser are you using?

Chrome

### What operating system are you using?

Windows

### How are you deploying Directus?

running locally

### LINKED ISSUE #10476: M2M items empty display field when creating item
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes my problem.

### Describe the Bug

When adding a new item which has m2m relation, the label is not visible. The label becomes visible when item is saved and is editable.

New entry:
image

Edit entry: 
image

Possible related to #9977 

### To Reproduce

Install v9.2.2. Create m2m relation. Add new item which has m2m relation.

### Errors Shown

Chrome logs `RangeError: Invalid time value` however not related (i think)

### What version of Directus are you using?

9.2.2

### What version of Node.js are you using?

directus docker image

### What database are you using?

postgres

### What browser are you using?

chrome

### What operating system are you using?

windows

### How are you deploying Directus?

docker

### LINKED ISSUE #10542: Upload Files in Item shows 2 added Files when uploading 1
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes my problem.

### Describe the Bug

UI shows 2 added files when uploading just one file.

image

When adding multiple files, this happens for every file.

### To Reproduce

image

I clicked on "Upload File" and chose 1 File to upload.

image

After saving it's just 1 File again.
image

### Errors Shown

none

### What version of Directus are you using?

9.2.2

### What version of Node.js are you using?

inside Docker: 16.13.1

### What database are you using?

Postgres 12

### What browser are you using?

Vivaldi

### What operating system are you using?

Windows

### How are you deploying Directus?

Docker

### LINKED ISSUE #10645: O2M field that targets a collection with a manual primary key (string) has display issues
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes my problem.

### Describe the Bug

Possibly related to #3834

When adding new `keys` to the `keychain`, they don't get displayed but are added in the background. The new `keys` will only be displayed after saving the `keychain`.

I know this explanation might be confusing but see reproduction below.

### To Reproduce

Create 2 collections:

1. `keys` collection with manually entered string as primary key called `key`. (everything else at default)
2. `keychains` collection (default values)

Go to `keychains` collection and create a new `O2M` field targeting the `keys` collection, create a new field (foreign key) during that step called `keychain_id` in the `keys` collection. Make sure to also do the following settings:

- Disable `select button`
- Set all relational triggers to delete (cascade)
- Setup Interface and Display to show the `key` value

Reason for these options: The `keys` collection is supposed to contain unique keys that are attached to one `keychain`.

Now visit the `keychains` collection and start creating a new entry. Add a few keys to the keychain and you will see the issue happening.

The issue only occurs when adding new entries, not when selecting existing entries.

Short and sweet video:

https://user-images.githubusercontent.com/7304212/147168654-2be14110-6d31-4571-b814-08331ef84a7d.mp4

### Errors Shown

N/A

### What version of Directus are you using?

9.3.0

### What version of Node.js are you using?

16.13.1

### What database are you using?

MariaDB 10.3.32

### What browser are you using?

Firefox Latest

### What operating system are you using?

Windows 10

### How are you deploying Directus?

Ubuntu Server 20.04, npm installation with a systemd service running it

### LINKED ISSUE #10646: O2M field does not check if manual primary key (string) already exists
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes my problem.

### Describe the Bug

This issue is connected to issue #10645 please see the collection setup there, I will use the collections from that issue to demonstrate the next very dangerous bug:

If you add a new `key` to a `keychain` but use a name that already exists in the `keys` collection you will essentially **STEAL** this `key` from the `keychain` it was previously attached to. This can result in field checks breaking if e. g. the field was configured to be `required`!

### To Reproduce

1. Create a new `keychain` in the `keychains` collection.
2. Add a `key` that already exists.
3. You will notice that the bug from issue #10645 did not apply here, since the key was already existing.
4. If you save now you **steal** this `key` from the other `keychain`.

Video showing the issue (I also reconfigure the field here to be `required` to show field checks breaking):

https://user-images.githubusercontent.com/7304212/147169523-0404dcd2-ac68-478b-bd18-23dd9fe8e6a4.mp4

### Errors Shown

N/A

### What version of Directus are you using?

9.3.0

### What version of Node.js are you using?

16.13.1

### What database are you using?

MariaDB 10.3.32

### What browser are you using?

Firefox Latest

### What operating system are you using?

Windows 10

### How are you deploying Directus?

Ubuntu Server 20.04, npm installation with a systemd service running it

### LINKED ISSUE #10916: M2M relationship field bugged (blank) when adding existing item, but fixed on save.
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes my problem.

### Describe the Bug

This bug happened after upgrading from 9.3 -> 9.4.x, it used to work on 9.3. On a M2M relationship, adding an existing item shows a blank item in the field. After saving, the item is correctly displayed.

Field before adding item:
image

Field after adding existing item:
image

Field after saving (properly displayed):
image

I have tried redefining the Display Template but the error persists. 

I asked a few times in the Discord if anyone else saw this error, finally had someone else confirm this bug. I haven't seen any other posts about it so I thought I would try creating a report. 

### To Reproduce

- Create a M2M relationship on 9.4.x and use a field from the related item in the display template.
- Use the 'add existing' button to add an item, the field will be bugged until save

### Errors Shown

There are no bugs shown in the browser console or the Directus server logs.

There also isn't a request shown in the Directus server logs on item select. Not sure if a request should be firing or not.

### What version of Directus are you using?

9.4.2

### What version of Node.js are you using?

v16.13.0

### What database are you using?

Postgres 12

### What browser are you using?

Chrome / Firefox

### What operating system are you using?

Windows 10

### How are you deploying Directus?

Self install on Digital Ocean droplet with a managed DB instance

### PR DESCRIPTION
**Pull request by Nitwel**
Rework every single relational interface.

closes #10645
closes #3834
closes #10646
closes #3889
closes #5540
closes #5752
closes #10476
closes #10542
closes #10916
closes #6085

---

Left to-do:

- [x] Confirm other relational interface issues
- [x] Implement a "where not in" syntax for o2m filters[^1] to finish m2m app filter

[^1]: So you can do `?filter[related_items][_none][category][_eq]=published`

### CODE REVIEW

**Comment by rijkvanzanten** [comment]
@Nitwel Lets do the various interfaces in different PRs 👍🏻 We don't have to replace all of them at the same time, and it means we can review each one individually a little better

**Comment by LasseRosenow** [comment]
When the treeview gets reworked, is there any chance to also add the option for it to work on recursive collections that are embedded inside a single instance o2m relation? This would give us to build real tree view menus that would still consist only of one collection.

Right now when I build a menu I can only see the first layer in the single instance collection because the treeview is incompatible

**Comment by rijkvanzanten** [comment]
> When the treeview gets reworked, is there any chance to also add the option for it to work on recursive collections that are embedded inside a single instance o2m relation? This would give us the ability to build real tree view menus that would still consist only of one collection.
> 
> Right now when I build a menu I can only see the first layer in the single instance collection because the treeview is incompatible

This sprint is primarily focussed on creating a new shared engine for the existing relational interfaces, and rework those to have feature parity (and solve a ton of bugs in the meantime). 

I can't make any promises around new features that will be added during this, as that's not the primary intention of this sprint 👍🏻

File: app/src/interfaces/list-o2m/list-o2m.vue [hunk 1]
```diff
@@ -1,515 +1,335 @@
 <template>
```
Comment (rijkvanzanten) [inline app/src/interfaces/list-o2m/list-o2m.vue:1]: Seems like `totalCount` is `"100"` for a list-o2m with a single item, not sure why yet

Comment (rijkvanzanten) [reply app/src/interfaces/list-o2m/list-o2m.vue:1]: I know: `existingItemCount.value` is a string (`1`), and the logic is `"1" + 0 + 0` = `"100"`

File: app/src/interfaces/list-o2m/list-o2m.vue [hunk 2]
```diff
@@ -1,515 +1,335 @@
...
-	<v-notice v-if="!relation" type="warning">
+	<v-notice v-if="!relationInfo" type="warning">
```
[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply app/src/interfaces/list-o2m/list-o2m.vue:2]: Select existing on a fresh load, deselect it, select it again 💥

File: app/src/interfaces/list-o2m/list-o2m.vue (hunk 1 above)
Comment (Nitwel) [reply app/src/interfaces/list-o2m/list-o2m.vue:1]: Haha, love those javascript pearls. 😄

File: app/src/interfaces/list-o2m/list-o2m.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply app/src/interfaces/list-o2m/list-o2m.vue:2]: - [ ] Lets right-align the pagination:

CleanShot 2022-04-06 at 17 15 22@2x

[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply app/src/interfaces/list-o2m/list-o2m.vue:2]: - [ ] Inconsistent "mark as deleted" behavior

Right now, rows are sometimes marked as deleted (with the red style), while other times, they're removed immediately. I believe this has something to do with deleting changes you've made vs deselecting records that are already saved to the database, but it feels a little janky in use.

I'm wondering if we should just hide the "marked as deleted" ones to make it consistent.

https://user-images.githubusercontent.com/9141017/162073107-3e22e925-e0cd-4f1a-a281-66fb17f7635d.mp4

[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply app/src/interfaces/list-o2m/list-o2m.vue:2]: - [ ] First load jumps

It doesn't know the length yet, so the skeleton loaders default to 0. We should default that to 1 so it doesn't jump (as much) on first load

https://user-images.githubusercontent.com/9141017/162073372-59eb2fe2-73ec-4246-bc1f-48242e9c9ab0.mp4

File: app/src/interfaces/list-m2m/list-m2m.vue [hunk 3]
```diff
@@ -1,282 +1,373 @@
 <template>
```
Comment (rijkvanzanten) [inline app/src/interfaces/list-m2m/list-m2m.vue:1]: ## Many to Many Feedback

Comment (rijkvanzanten) [reply app/src/interfaces/list-m2m/list-m2m.vue:1]: - [ ] Pagination alignment (https://github.com/directus/directus/pull/12082#discussion_r844428900)

Comment (rijkvanzanten) [reply app/src/interfaces/list-m2m/list-m2m.vue:1]: - [ ] Same selection bug as https://github.com/directus/directus/pull/12082#discussion_r844426830

**Comment by rijkvanzanten** [comment]
(I believe most of the above applies to every list style interface, as they all share the same underlying logic, but i've marked them where i ran into the issue initially)

File: app/src/interfaces/list-m2m/list-m2m.vue (hunk 3 above)
Comment (rijkvanzanten) [reply app/src/interfaces/list-m2m/list-m2m.vue:1]: - [ ] Nested filter doesn't filter out correct item

https://user-images.githubusercontent.com/9141017/162075799-30e8c069-f8c5-4840-979e-b481801d3795.mp4

(`Testing` should've been filtered out, as it's already selected)

Comment (rijkvanzanten) [reply app/src/interfaces/list-m2m/list-m2m.vue:1]: - [ ] Can't deselect nested selected item

https://user-images.githubusercontent.com/9141017/162077078-a97c4602-f433-4740-8b5f-53084bb272f0.mp4

**Comment by rijkvanzanten** [comment]
- [ ] Nested o2m in m2m saves changes to every item

https://user-images.githubusercontent.com/9141017/162296795-2979e3c6-6a5b-4b24-baab-b44efe745435.mp4

**Comment by rijkvanzanten** [comment]
- [ ] Not all existing M2M are showing up

https://user-images.githubusercontent.com/9141017/162298820-349b05b3-3cb5-4db6-b394-ab952e6e5739.mp4

(no custom filter configured)

**Comment by Nitwel** [comment]
The last one isn't a bug, we just haven't implemented `Implement a "where not in" syntax for o2m filters to finish m2m app filter` yet so the current filter doesn't make any sense.

**Review by rijkvanzanten — APPROVED** [review]

**Pull request merged**